File: Python/Algorithms/Sorting/ShellSort.py
# Python Program for Shell sort, including Knuth sequence for selecting gaps
# Shell Sort Algorithm: https://en.wikipedia.org/wiki/Shellsort
from random import *

def ShellSort(arr):
    gap = nextGap(len(arr))
    while gap > 0:
        for i in range(gap, len(arr)):

            # we need a temp variable to temporarily store the value before any swapping takes place
            temp = arr[i]

            # shift earlier gap-sorted elements up until the correct location for arr[i] is found
            j = i
            while j >= gap and arr[j-gap] > temp:
                arr[j] = arr[j-gap]
                j -= gap
            
            # put the temp (arr[i] originally) into it's correct location
            arr[j] = temp

        # now grab next gap value until we get down to 0
        print("array after shell gap value of %d:" %gap)
        printList(arr)
        gap = nextGap(gap)

# This method selects what the next gap value is going to be for sorting, using the Knuth Sequence formula.
def nextGap(previousGapValue):
    # Knuth gap sequence is set with gap = 3gap + 1 upto gap < data.length
    # So to reverse that direction (start with largest gap and whittle down) we return the previous gap value minus one divided by 3
    if (previousGapValue == 1):
        return 0
    elif((previousGapValue - 1)/3 <= 1):
        return 1
    else:
        return int(round((previousGapValue - 1)/3))


# I got sick of writing a bunch of loops to print out the list in a (relatively) readable format, so this method does it for you!
def printList(arr):
    n = len(arr)
    valueList = ""
    for i in range(n):
        valueList = valueList + "%d, " % arr[i]
    print(valueList)

# Code to test out sort
arr = []

# You can replace the number of values with anything, I find 98 is a good one to start because then you get to see gaps of 32, 10, 3 and the final with a gap of 1.
for i in range(98):
    arr.append(randint(1, 100))

print("Array before sorting")
printList(arr)
ShellSort(arr)
print("Array after sorting")
printList(arr)

File: Java/Algorithms/Sorting/ShellSort.java
import java.util.Scanner;
import java.util.Random;

public class ShellSort {

    // Normally best practice would be to have these methods part of a separate class so that they aren't static, but I'm lazy.
    public static void shellSort(int[] array)
    {
        int gap = nextGap(array.length);
        int temp, j;
        while(gap > 0)
        {
            for(int i = gap; i < array.length; i++)
            {
                // we need a temp variable to store the value before any swapping takes place.
                temp = array[i];

                // shift earlier gap-sorted elements up until the correct location for array[i] is found
                j = i;
                while(j >= gap && array[j-gap] > temp)
                {
                    array[j] = array[j-gap];
                    j -= gap;
                }

                // Now put the temp value into it's correct location
                array[j] = temp;

                
            }
            System.out.println("Array after sorting with a gap value of " + gap + " :");
            printList(array);

            // Grab the next gap value until we get to 0
            int newGap = nextGap(gap);
            gap = newGap;
        }
        
    }

    public static int nextGap(int previousGapValue)
    {
        if (previousGapValue == 1)
            return 0;
        else if ((previousGapValue - 1)/3 <= 1)
            return 1;
        else
            return Math.round((previousGapValue - 1)/3);
    }



    public static void printList(int[] array)
    {
        String result = "";
        for(int i = 0; i < array.length; i++)
        {
            result += array[i] + ", ";
        }
        System.out.println(result);
    }
    
    public static void main(String[] args)
    {
        Scanner input = new Scanner(System.in);
        Random random = new Random(); 

        System.out.println("Enter array size: ");
        int arrSize = input.nextInt();
        int[] numArray = new int[arrSize];

        System.out.println("Enter Max number in array");
        int maxValue = input.nextInt();

        for(int i = 0; i < arrSize; i++)
        {
            int value = random.nextInt(maxValue);
            numArray[i] = value;
        }

        System.out.println("Array before sorting: ");
        printList(numArray);

        shellSort(numArray);

        input.close();
    }

    
}